### ISSUE TITLE
/set_time command

### ISSUE THREAD

**Issue by Nisc3d** [manicdigger/manicdigger#107]
Maybe add a set_time command.
Syntax: /set_time (0:6:12:18)
The numbers are the times in the default.cs.

**Comment by croxxx** [comment]
Inplementing this needs a complete overhaul of how time is handled in MD currently.
- Adaption of SunMoonRenderer
- Adaption of time saved in savegames
- *eventually more I didn't think of right now...*

**Comment by croxxx** [comment]
Current time is static based on time since creation of world (taken from Server.cs):
```csharp
public class GameTime
{
    public long Ticks;
    public int TicksPerSecond = 64;
    public double GameYearRealHours = 24;
    public double GameDayRealHours = 1;
    public double HourTotal // 0 - 23
    {
        get
        {
            return ((double)Ticks / TicksPerSecond) / (GameDayRealHours * 60 * 60);
        }
    }
    public double YearTotal
    {
        get
        {
            return ((double)Ticks / TicksPerSecond) / (GameYearRealHours * 60 * 60);
        }
    }
}
```
This is also very closely related to server simulation time. It should be independent, though as game time might be variable.

@exemd and I spoke about this some time ago.
The discussion ended up on an interesting concept of chunk versioning instead of absolute time, but that's another issue. :grinning: 

Current system isn't fully functional anyway. If changing duration of days things get buggy.

**Comment by TobiMob** [comment]
I created [this branch](https://github.com/TobiMob/manicdigger_TimeOfDay) with a prototype for a more flexible TimeOfDay system.

Only Server.cs is changed and those two variables are the center of it:
```cs
    //current time of day
    private TimeSpan _tsIngameTime = new TimeSpan(9, 0, 0);// 9:00
  
    //This changes how fast time goes on
    //0 freezes the time
    //can be negative to let time run backwards
    private int _nIngameSecondsEveryRealTimeSecond = 600;
```

Changes:
--
**_tsIngameTime as the current TimeOfDay**
Decouples the ServerTicks from the IngameTime, 
to update the time independent from the Server Ticks.
A [TimeSpan](https://msdn.microsoft.com/en-us/library/System.TimeSpan(v=vs.110).aspx) can hold ~10,600,000 days (~29,000 years)

-Register SetTime-Command
-Change _tsIngameTime-Variable
-Profit.

--
**_nIngameSecondsEveryRealTimeSecond int defines how fast time goes by.**
The SunMoonRenderer works pretty solid.
So we can let the time run forward, backward or freeze it.

_nIngameSecondsEveryRealTimeSecond with a value of 600 still looks good for me,
but as the value increases, the sun starts to jump backwards.
Looks like some sort of interpolation is needed, to make faster clocks look smooth.

--
**Server Ticks are no longer saved, instead TimeOfDay is saved.**
(Discuss please)
Is there still any need to save the server-ticks?

----

This changes only affect the Server. It should also work with older clients.
I think _tsIngameTime and _nIngameSecondsEveryRealTimeSecond are a good way to go.

Any additions or concerns about this direction?

**Comment by croxxx** [comment]
Server Ticks in chunks is used to determine when a chunk last changed, I think. So changing this maybe requires some extra work.

SunMoonRenderer already handles such stuff? Good.
- Does changing time during runtime work with this?
- Can you change speed of time passing without having to reconnect?
- What happens if client gets disconnected from server? Does time still pass for them?

Personally, I'd like a system that repeats each day instead of a continuous one more. But that's my opinion.

So far: **good concept**

**Comment by TobiMob** [comment]
Most of the stuff was already implemented.
As far as I understand, it works currently like this:

If one of these occurs:
- Client connects
- TimeOfDay changes to a new sunlevel (15 ingame minutes)

then:
Server.NotifySeason(clientID) is called.
Which sends a Packet_ServerSeason to the client.
The package contains the TimeOfDay and the "speed of time" (meh).
This information is forwarded to the SunMoonRenderer.
The SunMoonRenderer sets its internal clock to the given time and starts running with the given speed,
until noted otherwise.

So the only thing that misses is a NotifySeason() call after a "speed of time" changes..
And maybe NotifySeason() should be renamed to NotifyTime() or such.

**Comment by croxxx** [comment]
Sounds quite straightforward.
Renaming definitely is a good idea. It doesn't have that much to do with seasons :smile:

**Comment by TobiMob** [comment]
Doesn't seem to complicated yet :smiley:.

How about storing the days?
The passed days are already stored in the Timespan.
They could be used for seasons, a farming mod, three growth or whatever.

Possible server command:
Useful could be:
- set time relative
- set time absolute
- set time speed

Commands could start with a /time then have a operation and then a value

/time set (d.)hh:mm:ss
Set time to (d.) possible for days

/time add (-)(d.)hh:mm:ss 
add/subtract given time to the current time

/time speed x 
for every realtime second x ingame seconds pass.

**Comment by croxxx** [comment]
Good ideas for the command syntax. :+1: 
I also found out that while chunksseenTime[] values are assigned, they never seem to be used in server. Maybe they can simply be thrown out (referring to the ones in ClientOnServer class).

I don't know if it is affected, but *server.simulationcurrentframe* is used in ServerChunk class, too:
```csharp
[ProtoContract()]
public class ServerChunk
{
    [ProtoMember(1, IsRequired = false)]
    public byte[] dataOld;
    [ProtoMember(6, IsRequired = false)]
    public ushort[] data;
    [ProtoMember(2, IsRequired = false)]
    public long LastUpdate;
    [ProtoMember(3, IsRequired = false)]
    public bool IsPopulated;
    [ProtoMember(4, IsRequired = false)]
    public int LastChange;
    public bool DirtyForSaving;
    [ProtoMember(5, IsRequired = false)]
    public List<Monster> Monsters = new List<Monster>();
    [ProtoMember(7, IsRequired = false)]
    public int EntitiesCount;
    [ProtoMember(8, IsRequired = false)]
    public ServerEntity[] Entities;
}
```
Or is that time stuff completely independent then? Would certainly be cool.

**Comment by TobiMob** [comment]
I would like the changes to only affect: 
- Server.NotifySeason() [or Server.NotifyTime()]
- ModManager.GetCurrentHourTotal();
- ModManager.GetCurrentYearTotal();

that should be enough.
(Anything else that needs to be affected by the TimeOfDay?)

[ChunkSimulation](https://github.com/manicdigger/manicdigger/blob/master/ManicDiggerLib/Server/Systems/ChunksSimulation.cs) for example uses the Server.simulationcurrentframe, and I would like to keep it that way. Since it would screw up, if we stop the time.

- Server.NotifySeason() [Server.NotifyTime()]
Should be called depending on the TimeOfDay-Variable.

- ModManager.GetCurrentHourTotal()
Should return TimeOfDay.Hours

- ModManager.GetCurrentYearTotal();
Maybe a TimeOfDay.Days / DaysPerYear
(and add a ModManager.SetDaysPerYear())

This could be enough, to make setting the TimeOfDay possible (without breaking any existing stuff).

In that turn, I would like to mark the SetGameYearRealHours in the Modmanager as obsolete.
I think a Year calculation based on Days would be more straight forward.
It could be changed to call the new SetDaysPerYear() Method, to let existing mods still work (somehow).
```cs
//Sets the number of real hours that one ingame year takes
[Obsolete("use SetDaysPerYear() instead")]
SetGameYearRealHours(double hours);
{
   SetDaysPerYear(hours);
}

///Sets the number of ingame-days a ingame-year has 
SetDaysPerYear(double days)
```

**Comment by croxxx** [comment]
@TobiMob You're right, time should be separate.
I can't think of anything else that would need to be altered.

I like the idea of days per year.

Also we could simply remove the SetGameYearRealHours() method.
The only place it is used in is the game core mods.
Also I don't know any mod that uses it apart from one I wrote (which is basically a copy of the season change code).

**Comment by croxxx** [comment]
Implemented in 468d38e9d5c69d4d1676db900cf8f2c9f568807b and following commits.
Maybe still requires some work on command input formatting.
